# feature scaling 
# standartization 

# subtracting the mean and dividing by the standard deviation 

tranforming the data we are working with into a standard scale 

subtracting the mean and deviding by standard deviaton
thus, we have forced the features to appear similar. 
this will ensure our linear combinations trest the two variables equally. 

# Feature selection through Standardization

In [1]:
# For these lessons we will need NumPy, pandas, matplotlib and seaborn
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

# and of course the actual regression (machine learning) module
from sklearn.linear_model import LinearRegression

## Load the data

In [2]:
# Load the data from a .csv in the same folder
data = pd.read_csv('102MUL~1.CSV')

# Let's explore the top 5 rows of the df
data.head()

SAT  Rand 1,2,3   GPA
0  1714           1  2.40
1  1664           3  2.52
2  1760           3  2.54
3  1685           3  2.74
4  1693           2  2.83

In [3]:
data.describe()

SAT  Rand 1,2,3        GPA
count    84.000000   84.000000  84.000000
mean   1845.273810    2.059524   3.330238
std     104.530661    0.855192   0.271617
min    1634.000000    1.000000   2.400000
25%    1772.000000    1.000000   3.190000
50%    1846.000000    2.000000   3.380000
75%    1934.000000    3.000000   3.502500
max    2050.000000    3.000000   3.810000

## Create the multiple linear regression

### Declare the dependent and independent variables

In [4]:
# There are two independent variables: 'SAT' and 'Rand 1,2,3'
x = data[['SAT','Rand 1,2,3']]

# and a single dependent variable: 'GPA'
y = data['GPA']

### Regression itself

In [5]:
reg = LinearRegression()

In [6]:
reg.fit(x,y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [7]:
reg.coef_

array([ 0.00165354, -0.00826982])

In [8]:
reg.intercept_

0.29603261264909353

### calculating the R-squared

## reg.score(x,y) returns the R-squred of a linear regression

In [9]:
reg.score(x,y)

0.40668119528142815

## the adjusted r-squred steps on the r-squared and adjusts for the nnumber of variables included in the model

### formula for adjusted R^2

$R^2_{adj.} = 1 - (1-R^2)*\frac{n-1}{n-p-1}$

In [10]:
#  formula for adjusted R^2

#  $R^2_{adj.} = 1 - (1-R^2)*\frac{n-1}{n-p-1}$

# n = observations 
# p = predictors 

In [11]:
x.shape

(84, 2)

In [12]:
r2 = reg.score(x,y)
n = x.shape[0]
p = x.shape[1]

adjusted_r2 = 1-(1-r2)*(n-1)/(n-p-1)
adjusted_r2

0.39203134825134

## adj.R^2 < R^2, therefore one or more of the predictors have little or no explanatory power.

### how to detect the variables which are unneeded in a model?

## Feature selection 

### F-regression
### feature_selection.f_regression
### F-regression creates simple linear regressions of each feature and the dependent variable 

In [13]:
from sklearn.feature_selection import f_regression

In [14]:
f_regression(x,y)

(array([56.04804786,  0.17558437]), array([7.19951844e-11, 6.76291372e-01]))

0: F-statistics    
1: p-values

In [15]:
p_values = f_regression(x,y)[1]
p_values

array([7.19951844e-11, 6.76291372e-01])

## e-11 means *10^-11, or means / 10^11
## e-1 means *10^1, or means /10

In [16]:
p_values.round(3)

array([0.   , 0.676])

## the p-value of SAT is 0.000, the p-value of random123 is 0.676

### Creating a summary table 

In [17]:
reg_summary = pd.DataFrame(data=x.columns.values, columns=['Features'])
reg_summary

Features
0         SAT
1  Rand 1,2,3

In [18]:
reg_summary['Coefficients'] = reg.coef_
reg_summary['p-values'] = p_values.round(3)

In [19]:
reg_summary

Features  Coefficients  p-values
0         SAT      0.001654     0.000
1  Rand 1,2,3     -0.008270     0.676

# p-values are one of the best ways yo determine if a variable is redundant, but they provide no information whatsoever about HOW USEFUL a variable is

# Multiple Linear Regression with sklearn - Exercise Solution

In [20]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

from sklearn.linear_model import LinearRegression

In [21]:
data = pd.read_csv('REAL_E~1.CSV')
data.head()

price     size  year
0  234314.144   643.09  2015
1  228581.528   656.22  2009
2  281626.336   487.29  2018
3  401255.608  1504.75  2015
4  458674.256  1275.46  2009

In [22]:
data.describe()

price         size         year
count     100.000000   100.000000   100.000000
mean   292289.470160   853.024200  2012.600000
std     77051.727525   297.941951     4.729021
min    154282.128000   479.750000  2006.000000
25%    234280.148000   643.330000  2009.000000
50%    280590.716000   696.405000  2015.000000
75%    335723.696000  1029.322500  2018.000000
max    500681.128000  1842.510000  2018.000000

In [23]:
x = data[['size', 'year']]
y = data['price']

In [24]:
reg = LinearRegression()
reg.fit(x,y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [25]:
reg.intercept_

-5772267.017463278

In [26]:
reg.coef_

array([ 227.70085401, 2916.78532684])

In [27]:
reg.score(x,y)

0.7764803683276796

In [28]:
def adj_r2(x,y):
    r2 = reg.score(x,y)
    n = x.shape[0]
    p = x.shape[1]
    adjusted_r2 = 1-(1-r2)*(n-1)/(n-p-1)
    return adjusted_r2

In [29]:
adj_r2(x,y)

0.7718717161282503

### It seems the the R-squared is only slightly larger than the Adjusted R-squared, implying that we were not penalized a lot for the inclusion of 2 independent variables.

In [30]:
reg.predict([[750, 2009]])

array([258330.34465995])

### Calculate the univariate p-values of the variables

In [31]:
from sklearn.feature_selection import f_regression

In [32]:
f_regression(x,y)

(array([285.92105192,   0.85525799]), array([8.12763222e-31, 3.57340758e-01]))

In [33]:
p_values = f_regression(x,y)[1]
p_values

array([8.12763222e-31, 3.57340758e-01])

In [34]:
p_values.round(3)

array([0.   , 0.357])

### Create a summary table with your findings

In [35]:
reg_summary = pd.DataFrame(data = x.columns.values, columns = ['Features'])
reg_summary ['Coefficients'] = reg.coef_
reg_summary ['p-values'] = p_values.round(3)
reg_summary

Features  Coefficients  p-values
0     size    227.700854     0.000
1     year   2916.785327     0.357

##### It seems that 'Year' is not event significant, therefore we should remove it from the model.

# Feature selection through Standardization

# standardization feature scaling - this translates to subtracting the mean and deviding by the standard deviation 

### we have forced the features to appear similar 

standardScaler() is a preprocessing module used to standardize (or scale) data 

## Standardization
Full documentation: https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html

In [36]:
# Import the preprocessing module
# StandardScaler is one of the easiest and 'cleanest' way to preprocess your data 

from sklearn.preprocessing import StandardScaler

In [37]:
# Create a StandardScaler instance (an empty Standardscaler project)

# scale our data

scaler = StandardScaler()

In [38]:
# Fit the input data (x)
# Essentially we are calculating the mean and standard deviation feature-wise
# (the mean of 'SAT' and the standard deviation of 'SAT', 
# as well as the mean of 'Rand 1,2,3' and the standard deviation of 'Rand 1,2,3')

# calculate the mean and standard deviation oof each feature, 
# this information will be stored in the scaler object. 
# scaler contains all stanardization info 

# Scaling machanism 

scaler.fit(x)

C:\Users\Junkie Wu\Anaconda3\lib\site-packages\sklearn\preprocessing\data.py:645: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)


StandardScaler(copy=True, with_mean=True, with_std=True)

### StandardScaler.transform(x) transforms the unscaled inputs using the information contained in the scaler object (feature-wise)

#### scaler.fit(x)
#### x_scaled = scaler.transform(x)

In [39]:
# The actual scaling of the data is done through the method 'transform()'
# Let's store it in a new variable, named appropriately 

x_scaled = scaler.transform(x)

C:\Users\Junkie Wu\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  after removing the cwd from sys.path.


In [40]:
# The result is an ndarray
x_scaled

array([[-0.70816415,  0.51006137],
       [-0.66387316, -0.76509206],
       [-1.23371919,  1.14763808],
       [ 2.19844528,  0.51006137],
       [ 1.42498884, -0.76509206],
       [-0.937209  , -1.40266877],
       [-0.95171405,  0.51006137],
       [-0.78328682, -1.40266877],
       [-0.57603328,  1.14763808],
       [-0.53467702, -0.76509206],
       [ 0.69939906, -0.76509206],
       [ 3.33780001, -0.76509206],
       [-0.53467702,  0.51006137],
       [ 0.52699137,  1.14763808],
       [ 1.51100715, -1.40266877],
       [ 1.77668568, -1.40266877],
       [-0.54810263,  1.14763808],
       [-0.77276222, -1.40266877],
       [-0.58004747, -1.40266877],
       [ 0.58943055,  1.14763808],
       [-0.78365788,  0.51006137],
       [-1.02322731,  0.51006137],
       [ 1.19557293,  0.51006137],
       [-1.12884431,  0.51006137],
       [-1.10378093, -0.76509206],
       [ 0.84424715,  1.14763808],
       [-0.95171405,  1.14763808],
       [ 1.62279723,  0.51006137],
       [-0.58004747,

## Regression with scaled features

In [41]:
# Creating a regression works in the exact same way
reg = LinearRegression()

# We just need to specify that our inputs are the 'scaled inputs'
reg.fit(x_scaled,y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [42]:
reg.coef_

array([67501.57614152, 13724.39708231])

In [43]:
reg.intercept_

292289.4701599997

## Creating a summary table


In [44]:
# As usual we can try to arrange the information in a summary table
# Let's create a new data frame with the names of the features
reg_summary = pd.DataFrame([['Bias'],['SAT'],['Rand 1,2,3']], columns=['Features'])

# Then we create and fill a second column, called 'Weights' with the coefficients of the regression
# Since the standardized coefficients are called 'weights' in ML, this is a much better word choice for our case
# Note that even non-standardized coeff. are called 'weights' 
# but more often than not, when doing ML we perform some sort of scaling
reg_summary['Weights'] = reg.intercept_, reg.coef_[0], reg.coef_[1]

In [45]:
# Now we have a pretty clean summary, which can help us make an informed decision about the importance of each feature
reg_summary

Features        Weights
0        Bias  292289.470160
1         SAT   67501.576142
2  Rand 1,2,3   13724.397082

## Weights is the 'machine learning word' for coefficients
## the bigger the weight, the bigger the impact
## the closer a weight is to 0, the smaller the impact.

## intercept = bias 

### using these new naming conventions we can conveniently distinguish between a "regular coefficients summary table" and this new one with standardized coefficients(weights)

In [46]:
# For these lessons we will need NumPy, pandas, matplotlib and seaborn
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

# and of course the actual regression (machine learning) module
from sklearn.linear_model import LinearRegression

In [47]:
# Load the data from a .csv in the same folder
data = pd.read_csv('102MUL~1_exercise.CSV')

# Let's explore the top 5 rows of the df
data.head()

SAT  Rand 1,2,3   GPA
0  1714           1  2.40
1  1664           3  2.52
2  1760           3  2.54
3  1685           3  2.74
4  1693           2  2.83

In [48]:
data.describe()

SAT  Rand 1,2,3        GPA
count    84.000000   84.000000  84.000000
mean   1845.273810    2.059524   3.330238
std     104.530661    0.855192   0.271617
min    1634.000000    1.000000   2.400000
25%    1772.000000    1.000000   3.190000
50%    1846.000000    2.000000   3.380000
75%    1934.000000    3.000000   3.502500
max    2050.000000    3.000000   3.810000

In [49]:
# There are two independent variables: 'SAT' and 'Rand 1,2,3'
x = data[['SAT','Rand 1,2,3']]

# and a single dependent variable: 'GPA'
y = data['GPA']

In [50]:
# Import the preprocessing module
# StandardScaler is one of the easiest and 'cleanest' ways to preprocess your data
from sklearn.preprocessing import StandardScaler

In [51]:
# Create an instance of the StandardScaler class
scaler = StandardScaler()

In [52]:
# Fit the input data (x)
# Essentially we are calculating the mean and standard deviation feature-wise 
# (the mean of 'SAT' and the standard deviation of 'SAT', 
# as well as the mean of 'Rand 1,2,3' and the standard deviation of 'Rand 1,2,3')
scaler.fit(x)

C:\Users\Junkie Wu\Anaconda3\lib\site-packages\sklearn\preprocessing\data.py:645: DataConversionWarning: Data with input dtype int64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)


StandardScaler(copy=True, with_mean=True, with_std=True)

In [53]:
# The actual scaling of the data is done through the method 'transform()'
# Let's store it in a new variable, named appropriately
x_scaled = scaler.transform(x)

C:\Users\Junkie Wu\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversionWarning: Data with input dtype int64 were all converted to float64 by StandardScaler.
  This is separate from the ipykernel package so we can avoid doing imports until


In [54]:
# The result is an ndarray
x_scaled

array([[-1.26338288, -1.24637147],
       [-1.74458431,  1.10632974],
       [-0.82067757,  1.10632974],
       [-1.54247971,  1.10632974],
       [-1.46548748, -0.07002087],
       [-1.68684014, -1.24637147],
       [-0.78218146, -0.07002087],
       [-0.78218146, -1.24637147],
       [-0.51270866, -0.07002087],
       [ 0.04548499,  1.10632974],
       [-1.06127829,  1.10632974],
       [-0.67631715, -0.07002087],
       [-1.06127829, -1.24637147],
       [-1.28263094,  1.10632974],
       [-0.6955652 , -0.07002087],
       [ 0.25721362, -0.07002087],
       [-0.86879772,  1.10632974],
       [-1.64834403, -0.07002087],
       [-0.03150724,  1.10632974],
       [-0.57045283,  1.10632974],
       [-0.81105355,  1.10632974],
       [-1.18639066,  1.10632974],
       [-1.75420834,  1.10632974],
       [-1.52323165, -1.24637147],
       [ 1.23886453, -1.24637147],
       [-0.18549169, -1.24637147],
       [-0.5608288 , -1.24637147],
       [-0.23361183,  1.10632974],
       [ 1.68156984,

In [62]:
# Creating a regression works in the exact same way
reg = LinearRegression()

# We just need to specify that our inputs are the 'scaled inputs'
reg.fit(x_scaled,y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [56]:
# Let's see the coefficients
reg.coef_

array([ 0.17181389, -0.00703007])

In [57]:
# And the intercept
reg.intercept_

3.330238095238095

In [58]:
# As usual we can try to arrange the information in a summary table
# Let's create a new data frame with the names of the features
reg_summary = pd.DataFrame([['Bias'],['SAT'],['Rand 1,2,3']], columns=['Features'])

# Then we create and fill a second column, called 'Weights' with the coefficients of the regression
# Since the standardized coefficients are called 'weights' in ML, this is a much better word choice for our case
# Note that even non-standardized coeff. are called 'weights' 
# but more often than not, when doing ML we perform some sort of scaling
reg_summary['Weights'] = reg.intercept_, reg.coef_[0], reg.coef_[1]

In [59]:
# Now we have a pretty clean summary, which can help us make an informed decision about the importance of each feature
reg_summary

Features   Weights
0        Bias  3.330238
1         SAT  0.171814
2  Rand 1,2,3 -0.007030

In [60]:
# weights is a machine learning word for coefficients 
# the bigger the weight, the bigger the impact

In [61]:
# f_regression 
# f statistic for each of those regressions 

## random 123 barely contributes to our output

## Making predictions with the standardized coefficients (weights)

In [63]:
# For simplicity, let's crete a new dataframe with 2 *new* observations
new_data = pd.DataFrame(data=[[1700,2],[1800,1]],columns=['SAT','Rand 1,2,3'])
new_data

SAT  Rand 1,2,3
0  1700           2
1  1800           1

In [65]:
reg.predict(new_data)

array([295.39979563, 312.58821497])

In [67]:
new_data_scaled = scaler.transform(new_data)
new_data_scaled

C:\Users\Junkie Wu\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: DataConversionWarning: Data with input dtype int64 were all converted to float64 by StandardScaler.
  """Entry point for launching an IPython kernel.


array([[-1.39811928, -0.07002087],
       [-0.43571643, -1.24637147]])

In [71]:
# We can make a prediction for a whole dataframe (not a single value)
# Note that the output is very strange (different from mine)
reg.predict(new_data_scaled)

array([3.09051403, 3.26413803])

## What if we removed the 'Random 1,2,3' variable?

In [69]:
# Theory suggests that features with very small weights could be removed and the results should be identical
# Moreover, we proved in 2-3 different ways that 'Rand 1,2,3' is an irrelevant feature
# Let's create a simple linear regression (simple, because there is a single feature) without 'Rand 1,2,3'
reg_simple = LinearRegression()

# Once more, we must reshape the inputs into a matrix, otherwise we will get a compatibility error 
# Note that instead of standardizing again, I'll simply take only the first column of x
x_simple_matrix = x_scaled[:,0].reshape(-1,1)

# Finally, we fit the regression
reg_simple.fit(x_simple_matrix,y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [70]:
# In a similar manner to the cell before, we can predict only the first column of the scaled 'new data'
# Note that we also reshape it to be exactly the same as x
reg_simple.predict(new_data_scaled[:,0].reshape(-1,1))

array([3.08970998, 3.25527879])